pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -848,3 +737,175 @@ func v1beta1ToV1(v1beta1Obj *admissionregistrationv1beta1.ValidatingAdmissionPol
 
 	return obj, nil
 }
+
+func (r *ReconcileConstraintTemplate) generateCRD(ctx context.Context, ct *v1beta1.ConstraintTemplate, proposedCRD, currentCRD *apiextensionsv1.CustomResourceDefinition, status *statusv1beta1.ConstraintTemplatePodStatus, logger logr.Logger, generateVAP bool) error {
+	if !operations.IsAssigned(operations.Generate) {
+		return nil
+	}
+	var newCRD *apiextensionsv1.CustomResourceDefinition
+	if currentCRD == nil {
+		newCRD = proposedCRD.DeepCopy()
+	} else {
+		newCRD = currentCRD.DeepCopy()
+		newCRD.Spec = proposedCRD.Spec
+	}
+
+	if err := controllerutil.SetControllerReference(ct, newCRD, r.scheme); err != nil {
+		return err
+	}
+
+	if currentCRD == nil {
+		logger.Info("creating crd")
+		if err := r.Create(ctx, newCRD); err != nil {
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not create CRD", status, err)
+			return err
+		}
+	} else if !reflect.DeepEqual(newCRD, currentCRD) {
+		logger.Info("updating crd")
+		if err := r.Update(ctx, newCRD); err != nil {
+			err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not update CRD", status, err)
+			return err
+		}
+		if ct.Annotations != nil {
+			ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation] = ""
```

Comment:
Do we need to reset this annotation when the CRD needs an update? the CRD should already been created and known to the apiserver right? and therefore, no need to wait for the vapb creation/updates?

Reply:
Yeah, you are correct. I initially thought that we need to wait because CRD would be different and api-server needs time to cache it. But it really doesn't matter, because all VAP cares about if kind. I will update the PR.